File: 13.Roman to Integer/answers.py
# 142ms 71.93%
# 倒序, 代码干净
class Solution(object):
    def romanToInt(self, s):
        pairs = {'M':1000, 'D':500, 'C':100, 'L':50, 'X':10, 'V':5, 'I':1}
        sum, prev = 0, 'I'
        for cur in s[::-1]:
            sum, prev = sum - pairs[cur] if pairs[cur] < pairs[prev] else sum + d[cur], cur
        return sum
        

# 122ms 98.54%
# 不倒序, 性能提升
class Solution(object):
    def romanToInt(self, s):
        pairs = {'M':1000, 'D':500, 'C':100, 'L':50, 'X':10, 'V':5, 'I':1}
        sum, prev = 0, "M"
        for cur in s:
            if pairs[prev] < pairs[cur]: sum -= 2*pairs[prev] 
            sum = sum + pairs[cur]
            prev = cur
        return sum

File: 13.Roman to Integer/answers.swift
class Solution {
    func romanToInt(_ s: String) -> Int {
        var pairs = [
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000
            
        ]
        var sLen = s.characters.count
        var sArr = s.characters.map { String($0) }
        var sum = pairs[sArr[sLen-1]]
        if (sLen == 1) {
            return sum!
        }
        for i in 1...(sLen-1) {
            var cur:Int? = pairs[sArr[i]] 
            var prev:Int? = pairs[sArr[i-1]] 
            if prev! < cur! {
                sum! = sum! - prev!
            } else {
                sum! = sum! + prev!
            }
        }
        return sum!
    }
}